src/artifacts.rs

```diff
@@ -1,111 +1,204 @@
+    // fully independent.
+    async fn reopen(&self) -> Result<Self> {
+        AsyncFile::options()
+            .read(true)
+            .write(true)
+            .open(format!("/proc/self/fd/{}", self.as_raw_fd()))
+            .await
+            .map_err(|e| e.into())
     }
+}
 
-    async fn get_file_or_none(&self, filename: &str) -> Result<Option<AsyncFile>> {
-        let reader = self.get_or_none(filename).await?;
-        Ok(if let Some(reader) = reader {
-            Some(async_save_to_tempfile(reader).await?)
-        } else {
-            None
+// This is a little hack to let us pass some thing to an async callback such that:
+// - We can get the value back when the callback completes.
+// - ...without having to use references (because async callback lifetimes are
+//   really messy).
+// In particular, it's needed for save_with implementations to be able to flush
+// the file at the very end while still giving it to the callback.
+pub struct ScopedCell<T> {
+    value: Option<T>,
+    once: Arc<OnceLock<T>>,
+}
+
+impl<T> ScopedCell<T> {
+    pub async fn run<Ret, Fut: Future<Output = Ret> + Send, F: (FnOnce(Self) -> Fut) + Send>(
+        value: T,
+        func: F,
+    ) -> (T, Ret) {
+        let mut once = Arc::new(OnceLock::new());
+
+        let ret = func(Self {
+            value: Some(value),
+            once: once.clone(),
         })
+        .await;
+        let value = Arc::get_mut(&mut once)
+            .and_then(|once| once.take())
+            .expect("ScopedCell leaked");
+        (value, ret)
     }
+}
 
-    async fn get_file(&self, filename: &str) -> Result<AsyncFile> {
-        let reader = self.get(filename).await?;
-        async_save_to_tempfile(reader).await
+impl<T> Drop for ScopedCell<T> {
+    fn drop(&mut self) {
+        _ = self.once.set(self.value.take().unwrap());
     }
+}
 
-    async fn get_data_or_none(&self, filename: &str) -> Result<Option<Vec<u8>>> {
-        Ok(
-            if let Some(mut reader) = self.get_or_none(filename).await? {
-                let mut data = Vec::new();
-                reader.read_to_end(&mut data)?;
-                Some(data)
-            } else {
-                None
-            },
-        )
-    }
+impl<T> Deref for ScopedCell<T> {
+    type Target = T;
 
-    async fn get_data(&self, filename: &str) -> Result<Vec<u8>> {
-        let mut reader = self.get(filename).await?;
+    fn deref(&self) -> &Self::Target {
+        self.value.as_ref().unwrap()
+    }
+}
 
-        let mut data = Vec::new();
-        reader.read_to_end(&mut data)?;
-        Ok(data)
+impl<T> DerefMut for ScopedCell<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.value.as_mut().unwrap()
     }
 }
 
-#[cfg(test)]
-pub mod test_support {
-    use std::{collections::HashMap, sync::Arc};
+#[async_trait]
+pub trait ArtifactDirectory: Send + Sync {
+    async fn open(&self, path: impl AsRef<Utf8Path> + Send) -> Result<AsyncFile>;
```

Comment:
Should this actually return an async file?  Looking at how this is actually used it seems to purely be for the artifact upload implementation. So an impl AsyncSeek + AsyncRead would likely be better to avoid things form trying to write to it.

Reply:
the issue I hit was that uploading via reqwest wants to take *ownership* of the body, which is the original place a while back I had to use `try_clone()`...a method that's specific async files. I had an idea of making that part of a custom trait but it felt a bit over-done (though in retrospect, maybe not *that* overdone given how wild the artifacts code kinda got).